### DISCUSSION TITLE
Filtering Nested Relationship Attributes using has/any

### DISCUSSION THREAD

**Discussion by jordantshaw** [sqlalchemy/sqlalchemy#7132] [Usage Questions] [answered]
Hi All!

I am having trouble filtering on multiple nested relationship attributes.  Normally, to filter on a relationship property, you would do something like this.

```
# Use has() or any() depending on the relationship type
query.filter(
    Order.customer.has(Customer.name == 'Jon Smith')
)
```

But lets say now there are multiple levels of relationships, and your child attribute, has a child relationship, then how would you filter based off that 2nd level? For instance, in the example below, how could I filter orders, based off the CustomerAddress.zip field?  

```
class Order(Base):
    order_id = Column(Integer)

    customer = relationship('Customer', backref='order')


class Customer(Base):
    customer_id = Column(Integer)
    name = Column(String)

    address = relationship('CustomerAddress', backref='customer')


class CustomerAddress(Base):
    address_id = Column(Integer)
    zip = Column(Integer)
```

**Comment by zzzeek** [comment, marked as answer]
you can nest has() / any() calls and that's what you would do here, like `Order.customer.has(Customer.address.has(...))`.    it will make for some very nested SQL that likely doesn't optimize well on most databases, though, better to construct a SELECT...JOIN of all three tables.

**Reply by jordantshaw** [reply]
Thank you!
